src/client/datascience/data-viewing/data-wrangler/dataWrangler.ts

```diff
@@ -0,0 +1,821 @@
+            await this.globalMemento.update(PREFERRED_VIEWGROUP, this.webPanel?.viewColumn);
+        }
+        this._onDidChangeDataWranglerViewState.fire();
+    }
+
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    protected onMessage(message: string, payload: any) {
+        let handled = false;
+        switch (message) {
+            case DataWranglerMessages.SubmitCommand:
+                this.handleCommand(payload).ignoreErrors();
+                handled = true;
+                break;
+
+            case DataWranglerMessages.RefreshDataWrangler:
+                this.refreshData().ignoreErrors();
+                handled = true;
+                break;
+
+            case CssMessages.GetMonacoThemeRequest:
+                void this.handleMonacoThemeRequest(payload);
+                handled = true;
+                break;
+
+            default:
+                break;
+        }
+
+        if (!handled) {
+            // Some messages will be handled by DataViewer
+            super.onMessage(message, payload);
+        }
+    }
+
+    private addToHistory(newHistoryItem: IHistoryItem) {
+        this.historyList.push(newHistoryItem);
+        this.postMessage(DataWranglerMessages.UpdateHistoryList, this.historyList).ignoreErrors();
+    }
+
+    private getCode() {
+        return this.historyList.map((item) => item.code).join('\n');
+    }
+
+    private async exportToCsv(_req: undefined, currentVariableName: string) {
+        const notebook = (this.dataProvider as IJupyterVariableDataProvider).notebook;
+        const fileInfo = await this.applicationShell.showSaveDialog({
+            saveLabel: DataScience.dataWranglerSaveCsv(),
+            filters: { CSV: ['csv'] }
+        });
+        if (fileInfo) {
+            const code = `${currentVariableName}.to_csv(path_or_buf=r'${fileInfo.fsPath}', index=False)`;
+            await notebook?.execute(code, '', 0, uuid(), undefined, false);
+        }
+    }
+
+    private async generatePythonCode() {
+        var dataCleanCode = this.getCode();
+
+        const doc = await this.documentManager.openTextDocument({
+            language: PYTHON_LANGUAGE,
+            content: dataCleanCode
+        });
+
+        await this.documentManager.showTextDocument(doc, 1, true);
+    }
+
+    private async generateNotebook() {
+        const dataCleanCode = this.getCode();
+        const notebookEditor = await this.commandManager.executeCommand(Commands.CreateNewNotebook);
+        // eslint-disable-next-line @typescript-eslint/no-explicit-any
+        const blankCell = (notebookEditor as any).document.cellAt(0) as NotebookCell;
+        await updateCellCode(blankCell, dataCleanCode);
+    }
+
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    private async handleCommand(payload: { command: DataWranglerCommands; args: any }) {
+        const notebook = (this.dataProvider as IJupyterVariableDataProvider).notebook;
```

Comment:
curious, can we just store this value once and make sure it's not undefined? do we need to keep getting references and checking for undefined?

Reply:
I believe it changes when `updateWithNewVariable` is called